src/platforms/evdev-rs/src/lib.rs

```diff
@@ -1,25 +1,1404 @@
+    handle: Option<thread::JoinHandle<()>>,
+
+    /// The write end of the pipe used to wake the input thread.
+    wfd: Option<io::OwnedFd>,
+
+    /// The channel used to send non-libinput commands to the input thread.
+    tx: Option<mpsc::Sender<ThreadCommand>>,
+}
+
+impl PlatformRs {
+    /// Start the input platform.
+    ///
+    /// This method will spawn a thread to handle input events from both
+    /// libinput and Mir. The thread will run until `stop()` is called.
+    pub fn start(&mut self) {
+        println!("Starting the evdev-rs platform");
+
+        let bridge = self.bridge.clone();
+        let device_registry = self.device_registry.clone();
+        let (rfd, wfd) = nix::unistd::pipe().unwrap();
+        let (tx, rx) = create_thread_command_channel();
+
+        self.wfd = Some(wfd);
+        self.tx = Some(tx);
+        self.handle = Some(thread::spawn(move || {
+            PlatformRs::run(bridge, device_registry, rfd, rx);
+        }));
+    }
+
+    pub fn continue_after_config(&self) {}
+    pub fn pause_for_config(&self) {}
+
+    /// Stop the input platform.
+    ///
+    /// This method will signal the input thread to stop and wait for it to finish.
+    pub fn stop(&mut self) {
+        if let Some(tx) = &self.tx {
+            let _ = tx.send(ThreadCommand::Stop);
+            if let Some(wfd) = &self.wfd {
+                let _ = nix::unistd::write(wfd.as_fd(), &[0u8]);
+            }
+        }
+
+        if let Some(handle) = self.handle.take() {
+            let _ = handle.join();
+            self.handle = None;
+        }
+
+        if let Some(wfd) = self.wfd.take() {
+            drop(wfd);
+            self.wfd = None;
+        }
+
+        self.tx = None;
+    }
+
+    pub fn create_device_observer(&self) -> Box<DeviceObserverRs> {
+        Box::new(DeviceObserverRs::new())
+    }
+
+    fn create_input_device(&self, device_id: i32) -> Box<InputDeviceRs<'_>> {
+        Box::new(InputDeviceRs {
+            device_id: device_id,
+            wfd: self.wfd.as_ref().unwrap().as_fd(),
+            tx: self.tx.clone().unwrap(),
+        })
+    }
+
+    fn run(
+        bridge_locked: cxx::SharedPtr<ffi::PlatformBridgeC>,
+        device_registry: cxx::SharedPtr<ffi::InputDeviceRegistry>,
+        rfd: io::OwnedFd,
+        rx: std::sync::mpsc::Receiver<ThreadCommand>,
+    ) {
+        let mut state = LibinputLoopState {
+            libinput: input::Libinput::new_with_udev(LibinputInterfaceImpl {
+                bridge: bridge_locked.clone(),
+                fds: Vec::new(),
+            }),
+            known_devices: Vec::new(),
+            next_device_id: 0,
+            button_state: 0,
+            pointer_x: 0.0,
+            pointer_y: 0.0,
+            scroll_axis_x_accum: 0.0,
+            scroll_axis_y_accum: 0.0,
+            x_scroll_scale: 1.0,
+            y_scroll_scale: 1.0,
+        };
+
+        state.libinput.udev_assign_seat("seat0").unwrap();
```

Comment:
We don't do this anywhere else?

Reply:
There's another TODO here: what to do if `udev_assign_seat()` returns `Err`; the current behaviour will end up calling `abort()`, which is probably not what we want to do here.